Clean up text in Xenium notebook (#1048)
AlexsLemonade/training-modules: spatial/05-xenium_processing.Rmd

Reviewer: Just to clarify here: if we had not removed the OME files, would there be an image?

Author: There would be an image to open locally but there woud not be an image in the SpatialExperiment object.

Reviewer: If I am understanding the output here correctly, are the `altExps` not named by the kind of control they include? You have to look at the rownames contents?

If that is the case, I think you will want to specifically mention that. I might even modify the code to pull out the prefix from the rownames rather than printing the whole summary. Though you do get the dimensions out, which is nice... of course you could print that too. If you do make either of these modifications, I would remove the live part!

Author: The names of the altExps do correspond to the control names, but I wanted to show the dimensions (as in how many of each control is used) and the rownames for each of these to note that these are distinguishable from the real probes. I'll remove the live part but let me know if you feel strongly about including the print out of all the altExps. If so, I'll just print the names.

Reviewer: That makes sense, I was just confused by the names not being in the output. I think this will fix it? 
```suggestion
altExpNames(spe) |> 
 purrr::set_names() |>
 purrr::map(\(name) altExp(spe, name))
```

Reviewer: Or could you just do `altExps(spe)` to see them all?

Reviewer: For consistency (and nicer display), I would use `dplyr::count()` here too.

Reviewer: ```suggestion
Cells that are not present in the annotation file will be assigned as `NA`.
```

Reviewer: presumably you only need one of these two?
```suggestion
 ggplot2::scale_color_distiller(palette = "YlOrRd", direction = 1)
```

Rows are features (panel genes plus control categories) and columns are segmented cells.
It contains only transcripts that passed the default quality threshold (Q-Score ≥ 20).
The same data are provided in HDF5 (`.h5`), a Matrix Market directory, and Zarr formats, but we will we use the `.h5` version.

##### `cells.parquet` / `cells.csv.gz` (and `cells.zarr.zip`)

This file provides our spatial coordinates and column (cell) metadata.
It includes each cell's ID, centroid coordinates (`x_centroid`, `y_centroid`) in microns, cell and nucleus area, transcript counts, and the segmentation method used.

##### `transcripts.parquet` / `transcripts.csv.gz`

This file contains the individual decoded transcript molecules, each with an `(x, y, z)` location, the gene it decoded to, a quality score, and the cell it was assigned to (if any).
This subcellular, molecule-level table is what distinguishes Xenium from spot-based methods. 
It is somewhat similar to `tissue_positions_list.csv` in Visium, but at far higher resolution.

##### `cell_boundaries.parquet` and `nucleus_boundaries.parquet` (plus `.csv.gz` versions)

> The polygon vertices defining each cell's and each nucleus's segmentation outline which is useful for plotting and for any polygon-based QC metrics like cell area and shape.

Reviewer: I think we can be a bit less specific here: I don;t think we need to say "vertices" for example...
```suggestion
These files describe the polygons that define each cell's and each nucleus's segmentation outline.
These polygons are useful for plotting and for calculating QC metrics like cell area and shape.
```

Nucleus boundaries come from segmenting the DAPI image and cell boundaries come from expanding or otherwise extending those nuclei (see the segmentation note below).
The count matrix is summarized at the **cell** level so that transcripts are assigned to the cell polygons. 
This means that `cell_boundaries` correspond to the columns of our object (or `spots` in Visium), while `nucleus_boundaries` is supplementary information about the nuclear compartment.

##### `analysis/` and `analysis.zarr.zip`

> The instrument's automated secondary analysis (e.g., dimensionality reduction and clustering) produced on-instrument.

Reviewer: ```suggestion
These files contain automated secondary analysis (e.g., dimensionality reduction and clustering) that are produced on-instrument.
```

##### `metrics_summary.csv` and `analysis_summary.html`

Run-level QC summaries as a CSV and a HTML report which includes transcript quality, segmentation method breakdown, control rates, and similar diagnostics.

##### Morphology images (`morphology.ome.tif`, `morphology_mip.ome.tif`, `morphology_focus.ome.tif` / `morphology_focus/`)

The tissue morphology images, including the DAPI nuclear stain and any additional stains that might be present for multi-modal assays.
We don't include these here to save space, but they are what you would use to visually inspect or re-run segmentation.

### Import with `SpatialExperimentIO`

We'll import the counts and spatial information into a `SpatialExperiment` (SPE) object using the `SpatialExperimentIO` package.
Note that here we are not using `VisiumIO` as our dataset is from Xenium, which is distinct from the previous Visium data sets we have been working with.

The only argument the function strictly requires is the path to the output bundle directory.

> Below we also pass a few arguments explicitly, like the count matrix and per-cell metadata, just so we can see what files are required, but we are still using the defaults.

Reviewer: ```suggestion
In the code below, we set a few arguments explicitly, like the count matrix and per-cell metadata, just so we can see what files are required, but we are still using the defaults.
```

##### `experiment.xenium`

A JSON file of run-level metadata including the gene panel, pixel size, software versions, and file paths.
This is the entry-point file that most loaders read first.

##### `cell_feature_matrix.h5` (and `cell_feature_matrix/`, `cell_feature_matrix.zarr.zip`)

The cell-by-feature count matrix, analogous to the feature-barcode matrix from Cell Ranger/Space Ranger.
Rows are features (panel genes plus control categories) and columns are segmented cells.
It contains only transcripts that passed the default quality threshold (Q-Score ≥ 20).
The same data are provided in HDF5 (`.h5`), a Matrix Market directory, and Zarr formats, but we will we use the `.h5` version.

##### `cells.parquet` / `cells.csv.gz` (and `cells.zarr.zip`)

This file provides our spatial coordinates and column (cell) metadata.
It includes each cell's ID, centroid coordinates (`x_centroid`, `y_centroid`) in microns, cell and nucleus area, transcript counts, and the segmentation method used.

##### `transcripts.parquet` / `transcripts.csv.gz`

> This file contains the individual decoded transcript molecules, each with an `(x, y, z)` location, the gene it decoded to, a quality score, and the cell it was assigned to (if any).

Reviewer: Wait, `z`? Do they all have the same `z` for a given slide or are we doing stacked imaging?

Author: It's 3D imaging so we get a z-position for everything. My understanding is that this is mostly used for cell segmentation and then we use the x and y coordinates for positions of the punctum on the slide.

### A note on segmentation 

> Because Xenium reports data per cell, every count depends on cell segmentation which is a computational approach that draws a boundary around each cell and decides which transcripts belong to each cell.

Reviewer: ```suggestion
Because Xenium reports data per cell, every count depends on the cell segmentation step.
Segmentation employs a computational algorithm to identify a boundary for each cell and decides which transcripts belong to each cell.
```

> Segmentation quality therefore directly affects downstream results, and it remains an active area of research as no method handles every tissue perfectly.

Reviewer: Do we want to note what tissues/cells are particularly difficult (I'm thinking neurons and muscle, just as examples)?
```suggestion
Segmentation quality therefore directly affects downstream results, but no current method handles every tissue perfectly.
Improving segmentation remains an active area of research.
```

- **Multimodal cell segmentation:** When the multi-tissue stain mix is used, a cell boundary stain (identified through membrane markers such as ATP1A1/CD45/E-Cadherin) is used to identify the true cell membrane, and an interior stain (18S rRNA) is used alongside the DAPI stain to identify the nuclei. 
In this scenario, nuclear expansion is used as a fallback.

This particular dataset only included a DAPI stain so cell segmentation was done using nuclear expansion. 

### A note on control features and `altExps`

A Xenium panel includes probes for both targeted genes and control probes. 

To understand the control categories, it helps to know how Xenium turns images into counts.
Each targeted gene is detected not by sequencing but by a characteristic pattern of fluorescent signals across imaging channels and cycles.
That pattern is called a **codeword**, and the full lookup table mapping codewords to genes is the **codebook** ([10x Genomics Xenium panel design concepts and terms](https://www.10xgenomics.com/support/software/xenium-panel-designer/latest/tutorials/xenium-panel-design-concepts-terms)).

> Decoding is the process of reading the observed signal at each detected fluorescent punctum and matching it to a codeword (and therefore a gene), with a quality score attached to each call.

Reviewer: I think you should define punctum. I might introduce it before the codewords? Maybe something as simple as: "Each targeted gene is detected not by sequencing but by a characteristic pattern of fluorescent signals at a specific site (punctum) across imaging channels and cycles."

The codebook deliberately reserves some codewords and probes as controls, which let us estimate background and error rates:

- **Negative control probes** are physical probes included in the panel that target sequences _not_ present in the transcriptome. Counts on these reflect nonspecific probe binding, so they measure assay background.

> - **Negative control codewords** are valid codewords that are deliberately left unassigned to any real gene. 

Reviewer: ```suggestion
- **Negative control codewords** are valid codewords that are deliberately left unassigned to any real probe in the panel. 
```

Counts on these reflect optical or decoding errors, so they measure decoding specificity.
- **Antisense** are probes targeting the antisense strand. 
Counts on these reflect nonspecific probe binding and measure assay background. 

> - **BLANK** codewords are empty codewords that map to nothing. 

Reviewer: Okay, maybe I am misunderstanding the negative control codewords; I don't think I understand the difference between those and blank?

Author: So my understanding of these which I am getting through the [10x documentation on controls](https://www.10xgenomics.com/support/software/xenium-onboard-analysis/latest/algorithms-overview/xoa-algorithms) and some googling is that negative control codewords are real codewords that don't correspond to a gene or probe that's included in the panel. The BLANK ones (which also are referred to as Unassigned) are codewords that are totally unused by 10x and don't correspond to any genes. I'm going to adjust the text to hopefully clarify this.

These are also used to identify decoding specificity. 
Note that this category does not appear in newer versions and has been replaced with Unassigned codewords.

Because a high-quality cell should have essentially no control counts, these features are a powerful QC signal and we will use them below.

> To keep them from being mistaken for real gene counts, `readXeniumSXE()` moves them out of the main matrix and into alternative experiments (`altExps`) stored alongside the main object.

Reviewer: ```suggestion
To keep control counts from being mistaken for real gene counts, `readXeniumSXE()` moves them out of the main matrix and into alternative experiments (`altExps`) stored alongside the main object.
```

  forcats::fct_relevel(names(cell_type_palette))
```

Now that we have prepared our data, we can create our plot showing the lumped cell type annotations across the slide. 

```{r plot-celltypes, live = TRUE}
#| fig.width: 8
#| message: FALSE

# plot all spots colored by cell type 
ggspavis::plotCoords(spe, annotate = "cell_types_lumped") +
  scale_color_manual(values = cell_type_palette)
```

## Identifying low quality cells 

Cells from imaging based methods can be considered low-quality for one of two reasons: 

1. The cell has a high level of background noise or has had issues decoding the codewords, as indicated by high per-cell counts for the negative control probes or negative codewords. 
For Xenium, this information is stored in `control_probe_counts` and `control_codeword_counts` in the `colData`. 
2. The cell has a low number of gene expression counts compared to other cells of the same size, indicating the cells may be debris or segmentation artifacts. 
For Xenium, this information is stored in `total_counts` and `cell_area` in the `colData`. 

**Note:** For other imaging based methods, this information may not be pre-calculated. 

> In this case the [`SpaceTrooper`](https://bioconductor.org/packages/SpaceTrooper) ([Banzi _et al.,_ 2025](https://doi.org/10.64898/2025.12.24.696336)) package can be used to compute per-cell quality-control metrics that are tailored to imaging-based spatial data. 

Reviewer: Bigger link target.
```suggestion
In this case the [`SpaceTrooper` package](https://bioconductor.org/packages/SpaceTrooper) ([Banzi _et al.,_ 2025](https://doi.org/10.64898/2025.12.24.696336)) can be used to compute per-cell quality-control metrics that are tailored to imaging-based spatial data. 
```

```

## Identifying low quality cells 

Cells from imaging based methods can be considered low-quality for one of two reasons: 

1. The cell has a high level of background noise or has had issues decoding the codewords, as indicated by high per-cell counts for the negative control probes or negative codewords. 

> For Xenium, this information is stored in `control_probe_counts` and `control_codeword_counts` in the `colData`. 

Reviewer: Were these calculated at import, or already by Space Ranger, or is this something that will be calculated by `SpaceTrooper`?

Reviewer: Update reading on: it seems these must already exist since you set `use_altexps = FALSE`. I still think it is worth clarifying when they were calculated though.

2. The cell has a low number of gene expression counts compared to other cells of the same size, indicating the cells may be debris or segmentation artifacts. 
This information should be calculated for each cell. 

We will use [`SpaceTrooper`](https://bioconductor.org/packages/SpaceTrooper) ([Banzi _et al.,_ 2025](https://doi.org/10.64898/2025.12.24.696336)) to compute per-cell quality-control metrics that are tailored to imaging-based spatial data. 

> This includes cell area and total counts which is calculated from the `counts` assay in the main experiment. 

Reviewer: I'm not sure this sentence is actually needed? I feel like it can wait until you do the calculations and look at the output?
```suggestion
This includes cell area and total counts, which is calculated from the `counts` assay in the main experiment. 
```

# note that some of this may have been pre-calculated for this experiment but we'll re-do that here
spe <- SpaceTrooper::spatialPerCellQC(
  spe,
  use_altexps = FALSE # don't calculate metrics for altexperiments
)

head(colData(spe))
```

`spatialPerCellQC()` adds several per-cell QC columns to `colData()`. 
The ones we will use for QC include `total` and `Area_um`: 

- `total`: total transcript counts per cell summed over all panel genes.
- `Area_um`: the cell's area in square microns, derived from its segmentation polygon.

> **Note:** For this particular experiment, these metrics were already stored in the columns `total_counts` and `cell_area` found in the `colData`, but this will not be the case for all imaging-based experiments. 

Reviewer: Is it true for all Xenium data? If so, I wonder if we might want to skip the `SpaceTrooper` calculations (despite the name) and just use the existing stats. We can still mention that it might be useful for other technologies to get some of the same info, but if it isn't actually calculating anything new for us, I question whether we should include it here.

Reviewer: Update after looking a bit more; it seems like it also adds `ctrl_total_ratio`? Is there a reason we are not using that as well?

Author: I think something is off with the implementation of `SpaceTrooper` with `Xenium` data since the control counts that come out of `SpaceTrooper` are all 0 for this experiment and that means the `ctrl_total_ratio` is all 0 and pretty useless. 

I also see that `cell_area` and `total_counts` are standard output from [Xenium on the most recent version](https://www.10xgenomics.com/support/software/xenium-onboard-analysis/latest/analysis/xoa-output-understanding-outputs#cell-summary), so I think we should just use that and sad a sad farewell to `SpaceTrooper`. I'm going to edit this section accordingly.

p1 + p2
```

To flag these cells, we summarize each cell by its counts-per-area ratio and look for cells that sit far below the bulk of the distribution.
Specifically, we use `scater::isOutlier()` to flag cells whose ratio falls more than three median absolute deviations (MADs) below the median.

* The MAD is a robust measure of spread and is less sensitive to extreme values than the standard deviation. 

> * We will use the common cutoff of three MADs. 

Reviewer: I was confused by "three MADs" and wondered if we were combining three different estimates. 
This bullet point may not be needed, since you do say you will do this in the sentence before.
```suggestion
* We will use the common cutoff of 3 * MAD. 
```

* We only flag the low side, since it is unusually _few_ counts for a cell's area that signals debris or a failed segmentation.

> * We work on the log scale because the ratio is strictly positive and right-skewed, which makes a symmetric MAD rule more appropriate.

Reviewer: I am not sure what you are exactly saying here? (I think I understand the decision, but the sentence only talks about right skew and symmetric, which seem opposed)

> outliers

Reviewer: I don't think we want to print the `outliers` vector here. maybe `sum(outliers)`?

Author: I was printing this to show the structure of the output and that there's an attribute that stores the actual threshold that we want to use. I'll adjust this to use `str()` instead.

```

Now we can pull out the specific threshold corresponding to 3 MADs below the median from the `outliers` object. 

```{r define-threshold, live=TRUE}
# pull out the numeric threshold for plotting
threshold <- attr(outliers, "threshold")[1]
```

We'll also plot the distribution of the ratio and the chosen cutoff. 

```{r plot-threshold}
# visualize the distribution and the chosen threshold
ggplot(mapping = aes(log(count_area_ratio))) +
  geom_histogram(bins = 100, fill = "grey90", color = "black") +
  theme_bw() +
  geom_vline(xintercept = log(threshold)) +
  labs(
    x = "Log (total counts/area)"
  )
```

## Filtering low quality spots

Now we can filter by combining the counts-per-area QC metric with information from the control features as a secondary quality signal. 
A high-quality cell should have essentially no control counts. 
Any counts in these categories indicate background or errors in that cell. 

> Therefore we exclude cells with a nonzero count in either control category or with a counts-per-area ratio below the cutoff. 

Reviewer: Is there any value in using the `ctrl_total_ratio` rather than a count cutoff?

sum(exclude) / length(exclude)
```

> Now we can add a new column to the SPE indicating which cells to exclude and use that to visualize what cells will be excluded during filtering. 

Reviewer: Is it worth showing the cells flagged by the two measures separately? I think it may be?
